src/lib.rs

```diff
@@ -0,0 +1,534 @@
+    }
+}
+
+impl From<original::FftDirection> for rustfft::FftDirection {
+    fn from(direction: original::FftDirection) -> Self {
+        match direction {
+            original::FftDirection::Forward => rustfft::FftDirection::Forward,
+            original::FftDirection::Inverse => rustfft::FftDirection::Inverse,
+        }
+    }
+}
+
+pub struct FftPlanner(Mutex<original::FftPlanner<f64>>);
+
+impl FftPlanner {
+    pub fn new(state: original::FftPlanner<f64>) -> Self {
+        Self(Mutex::new(state))
+    }
+}
+
+impl rustfft::FftPlanner for FftPlanner {
+    fn new_fft_planner() -> Handle<FftPlanner> {
+        Handle::new(FftPlanner::new(original::FftPlanner::new()))
+    }
+    fn plan_fft(&self, len: u32, direction: rustfft::FftDirection) -> Handle<InnerFft> {
+        let mut planner = self.0.lock().expect("Mutex was poisioned");
+        Handle::new(InnerFft::new(
+            planner.plan_fft(len as usize, direction.into()),
+        ))
+    }
+
+    fn plan_fft_forward(&self, len: u32) -> wai_bindgen_rust::Handle<crate::InnerFft> {
+        let mut planner = self.0.lock().expect("Mutex was poisioned");
+        Handle::new(InnerFft::new(
+            planner.plan_fft(len as usize, original::FftDirection::Forward),
+        ))
+    }
+
+    fn plan_fft_inverse(&self, len: u32) -> wai_bindgen_rust::Handle<crate::InnerFft> {
+        let mut planner = self.0.lock().expect("Mutex was poisioned");
+        Handle::new(InnerFft::new(
+            planner.plan_fft(len as usize, original::FftDirection::Forward),
+        ))
+    }
+}
+
+pub trait Algo {
+    fn algo_len(&self) -> u32;
+    fn algo_fft_direction(&self) -> rustfft::FftDirection;
+    fn algo_process_with_scratch(
+        &self,
+        buffer: Vec<Handle<Complex>>,
+        scratch: Vec<Handle<Complex>>,
+    );
+    fn algo_process_outofplace_with_scratch(
+        &self,
+        input: Vec<Handle<Complex>>,
+        output: Vec<Handle<Complex>>,
+        scratch: Vec<Handle<Complex>>,
+    );
+    fn algo_get_inplace_scratch_len(&self) -> u32;
+    fn algo_get_outofplace_scratch_len(&self) -> u32;
+    fn algo_process(&self, buffer: Vec<Handle<Complex>>);
+}
+
+impl<T> Algo for T
+where
+    T: original::Fft<f64> + original::Direction + original::Length,
+{
+    fn algo_len(&self) -> u32 {
+        self.len() as u32
+    }
+
+    fn algo_fft_direction(&self) -> rustfft::FftDirection {
+        self.fft_direction().into()
+    }
+
+    fn algo_process_with_scratch(
+        &self,
+        buffer: Vec<Handle<Complex>>,
+        scratch: Vec<Handle<Complex>>,
+    ) {
+        let mut buffer_vec = buffer
+            .iter()
+            .map(|complex| complex.0)
+            .collect::<Vec<original::num_complex::Complex64>>();
+
+        let mut scratch_vec = scratch
+            .iter()
+            .map(|complex| complex.0)
+            .collect::<Vec<original::num_complex::Complex64>>();
+
+        self.process_with_scratch(&mut buffer_vec, &mut scratch_vec)
+    }
```

Comment:
This function doesn't actually do anything.

The purpose of the `xxx_with_scratch()` methods is so you can reuse the `scratch` buffer passed in. However, because WebAssembly doesn't let you have references to memory on the host, our `scratch` variable is passed in by value... which means any changes we make to it in this function will never be seen by the caller.

You should remove the function completely and just stick with `process()`.

Reply:
Removed all the scratch methods from here